piglow/piglow_test.go

```diff
@@ -0,0 +1,260 @@
+package piglow_test
+
+import (
+	"bytes"
+	"reflect"
+	"testing"
+
+	"github.com/goiot/devices/piglow"
+	"golang.org/x/exp/io/i2c/driver"
+)
+
+type opener struct {
+	Buffer *bytes.Buffer
+}
+
+func (o opener) Open() (driver.Conn, error) {
+	return &conn{
+		Buffer: o.Buffer,
+	}, nil
+}
+
+type conn struct {
+	Buffer *bytes.Buffer
+}
+
+func (c conn) Tx(w, r []byte) error {
+	if w != nil {
+		if _, err := c.Buffer.Write(w); err != nil {
+			return err
+		}
+	}
+
+	if r != nil {
+		if _, err := c.Buffer.Read(r); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (conn) Close() error {
+	return nil
+}
+
+func openPiGlow(t *testing.T) (*piglow.PiGlow, *bytes.Buffer) {
+	o := opener{
+		Buffer: bytes.NewBuffer([]byte{}),
+	}
+	device, err := piglow.Open(o)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	return device, o.Buffer
+}
+
+func assert(t *testing.T, expected, actual interface{}) {
+	if !reflect.DeepEqual(expected, actual) {
+		t.Fatal("expected", expected, "to equal", actual)
+	}
+}
+
+func TestGreen(t *testing.T) {
+	device, buffer := openPiGlow(t)
+	buffer.Reset()
+
+	if err := device.Green(1); err != nil {
+		t.Fatal(err)
+	}
+
+	expected := []byte{
+		0x04, 0x01,
+		0x06, 0x01,
+		0x0E, 0x01,
+		0x16, 0xFF,
+	}
+
+	assert(t, expected, buffer.Bytes())
+}
+
+func TestBlue(t *testing.T) {
+	device, buffer := openPiGlow(t)
+	buffer.Reset()
+
+	if err := device.Blue(1); err != nil {
+		t.Fatal(err)
+	}
+
+	expected := []byte{
+		0x05, 0x01,
+		0x0C, 0x01,
+		0x0F, 0x01,
+		0x16, 0xFF,
+	}
+
+	assert(t, expected, buffer.Bytes())
+}
+
+func TestYellow(t *testing.T) {
+	device, buffer := openPiGlow(t)
+	buffer.Reset()
+
+	if err := device.Yellow(1); err != nil {
+		t.Fatal(err)
+	}
+
+	expected := []byte{
+		0x03, 0x01,
+		0x09, 0x01,
+		0x010, 0x01,
+		0x16, 0xFF,
+	}
+
+	assert(t, expected, buffer.Bytes())
+}
+
+func TestOrange(t *testing.T) {
+	device, buffer := openPiGlow(t)
+	buffer.Reset()
+
+	if err := device.Orange(1); err != nil {
+		t.Fatal(err)
+	}
+
+	expected := []byte{
+		0x02, 0x01,
+		0x08, 0x01,
+		0x011, 0x01,
+		0x16, 0xFF,
+	}
+
+	assert(t, expected, buffer.Bytes())
+}
+
+func TestWhite(t *testing.T) {
```

Comment:
I struggled with turning this into a table test or not. Having them as separate tests makes them a little easier to read and follow. There's a clear mapping between the implementation and the test. This does however cause some boilerplate setup per function.